### PR TITLE
feat(test runner): allow stopping testrun with escape

### LINKED ISSUE #32579: [Internal]: Watch mode should have a way of stopping tests
Tests can take long to execute. All of our interfaces have some way of stoping a test run, watch mode should too.

### PR DESCRIPTION
**Pull request by Skn0tt**
Closes https://github.com/microsoft/playwright/issues/32579

### CODE REVIEW

**Comment by pavelfeldman** [comment]
What do you think about handling the first Ctrl+C (Cmd+C) to stop? It might be a bit more discoverable. Esc is designed to be used for escape sequences in the terminal environments, so there is an expectation that something follows Esc.

**Comment by pavelfeldman** [comment]
> We're listening to both Ctrl+C, Escape and their Control sequence equivalents. Do you think we should be listening exclusively to Ctrl+C?

I misunderstood the intention behind the patch.

File: packages/playwright/src/runner/watchMode.ts [hunk 1]
```diff
@@ -251,12 +252,37 @@ export async function runWatchModeLoop(configLocation: ConfigLocation, initialOp
...
 
+function readKeyPress(handler: (text: string, key: any) => void): Disposable {
+  const rl = readline.createInterface({ input: process.stdin, escapeCodeTimeout: 50 });
+  readline.emitKeypressEvents(process.stdin, rl);
+  if (process.stdin.isTTY)
+    process.stdin.setRawMode(true);
+
+  process.stdin.on('keypress', handler);
+
+  return {
```
Comment (pavelfeldman) [inline packages/playwright/src/runner/watchMode.ts:263]: returning the callback effect-style would also work. We only use Disposable when we interoperate with vscode.

File: packages/playwright/src/runner/watchMode.ts [hunk 2]
```diff
@@ -316,15 +338,9 @@ Change settings
...
-
-  process.stdin.on('keypress', handler);
-  void result.finally(() => {
-    process.stdin.off('keypress', handler);
-    rl.close();
-    if (process.stdin.isTTY)
-      process.stdin.setRawMode(false);
   });
+
+  void result.finally(reader.dispose);
```
Comment (pavelfeldman) [inline packages/playwright/src/runner/watchMode.ts:343]: I think readKeyPress should have this code, not readCommand

**Review by pavelfeldman — APPROVED** [review]

File: packages/playwright/src/runner/watchMode.ts (hunk 2 above)
Comment (Skn0tt) [reply packages/playwright/src/runner/watchMode.ts]: i've moved `readKeyPress` to return a promise in https://github.com/microsoft/playwright/pull/32584/commits/f0de6ccd81ff07f9e947625839b3335e0da3ba4a. That changes quite a bit about the code, but makes it easier to reason about in general - and it also moves the disposal into `readKeyPress` 👍

File: packages/playwright/src/runner/watchMode.ts (hunk 1 above)
Comment (Skn0tt) [reply packages/playwright/src/runner/watchMode.ts]: done in https://github.com/microsoft/playwright/pull/32584/commits/f0de6ccd81ff07f9e947625839b3335e0da3ba4a

**Comment by Skn0tt** [comment]
Unsure why this is failing on Windows. Need to investigate

**Review by pavelfeldman — APPROVED** [review]

**Pull request merged**